drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua

```diff
@@ -0,0 +1,1580 @@
+    DoorLock.attributes.NumberOfPINUsersSupported,
+    DoorLock.attributes.MaxPINCodeLength,
+    DoorLock.attributes.MinPINCodeLength,
+    DoorLock.attributes.RequirePINforRemoteOperation
+  },
+  [capabilities.lockSchedules.ID] = {
+    DoorLock.attributes.NumberOfWeekDaySchedulesSupportedPerUser,
+    DoorLock.attributes.NumberOfYearDaySchedulesSupportedPerUser
+  }
+}
+
+local subscribed_events = {
+  [capabilities.lock.ID] = {
+    DoorLock.events.LockOperation
+  },
+  [capabilities.lockAlarm.ID] = {
+    DoorLock.events.DoorLockAlarm
+  },
+  [capabilities.lockUsers.ID] = {
+    DoorLock.events.LockUserChange
+  }
+}
+
+local function is_new_matter_lock_products(opts, driver, device)
+  if device.network_type ~= device_lib.NETWORK_TYPE_MATTER then
+    return false
+  end
+  for _, p in ipairs(NEW_MATTER_LOCK_PRODUCTS) do
+    if device.manufacturer_info.vendor_id == p[1] and
+      device.manufacturer_info.product_id == p[2] then
+        return true
+    end
+  end
+  return false
+end
+
+local function find_default_endpoint(device, cluster)
+  local res = device.MATTER_DEFAULT_ENDPOINT
+  local eps = device:get_endpoints(cluster)
+  table.sort(eps)
+  for _, v in ipairs(eps) do
+    if v ~= 0 then --0 is the matter RootNode endpoint
+      return v
+    end
+  end
+  device.log.warn(string.format("Did not find default endpoint, will use endpoint %d instead", device.MATTER_DEFAULT_ENDPOINT))
+  return res
+end
+
+local function component_to_endpoint(device, component_name)
+  return find_default_endpoint(device, clusters.DoorLock.ID)
+end
+
+local function device_init(driver, device)
+  device:set_component_to_endpoint_fn(component_to_endpoint)
+  device:subscribe()
+ end
+
+local function device_added(driver, device)
+  device:emit_event(capabilities.lockAlarm.alarm.clear({state_change = true}))
+end
+
+local function do_configure(driver, device)
+  local user_eps = device:get_endpoints(DoorLock.ID, {feature_bitmap = DoorLock.types.Feature.USER})
+  local pin_eps = device:get_endpoints(DoorLock.ID, {feature_bitmap = DoorLock.types.Feature.PIN_CREDENTIAL})
+  local week_schedule_eps = device:get_endpoints(DoorLock.ID, {feature_bitmap = DoorLock.types.Feature.WEEK_DAY_ACCESS_SCHEDULES})
+  local year_schedule_eps = device:get_endpoints(DoorLock.ID, {feature_bitmap = DoorLock.types.Feature.YEAR_DAY_ACCESS_SCHEDULES})
+
+  local profile_name = "lock"
+  if #user_eps > 0 then
+    profile_name = profile_name .. "-user"
+    if #pin_eps > 0 then
+      profile_name = profile_name .. "-pin"
+    end
+    if #week_schedule_eps + #year_schedule_eps > 0 then
+      profile_name = profile_name .. "-schedule"
+    end
+  else
+    profile_name = "base-lock"
+  end
+  device.log.info(string.format("Updating device profile to %s.", profile_name))
+  device:try_update_metadata({profile = profile_name})
+end
+
+local function info_changed(driver, device, event, args)
+  for cap_id, attributes in pairs(subscribed_attributes) do
```

Comment:
This behavior for checking the driver subscribed attributes and adding them if the device supports the corresponding capability is already supported by the [populate_subscribed_attributes_from_driver_defaults() function](https://github.ecodesamsung.com/iot-hub/scripting-engine/blob/main/lua_libs/st/matter/device.lua#L224) that is called each time `device:subscribe()` is called, so just calling `device:subscribe()` in this case should be enough to handle this.

Also, you should check first that the profile changed before resubscribing, or else we'd end up just resubscribing to the same attributes/events as before:
```
if device.profile.id ~= args.old_st_store.profile.id then
  device:subscribe()
end
```

Reply:
@ctowns 
Thank you for your explanation~! I'll fix this part

Reply:
Thank you for your explanation. but when the above patch was applied, the driver did not work.
I think it is because main driver does not have some capabilities used by new-matter-lock driver. the subscribed attributes and event list in main driver is different with the list in sub driver.

Reply:
Thanks for trying that Hunsup! On further investigation it looks like the issue is that the `populate_subscribed_attributes_from_driver_defaults()` function only uses the subscribed attributes that are listed in the parent driver, not the sub driver. Since this is a sub driver, then the the updated `subscribed_attributes` that are defined in the `new-matter-lock` sub driver are not used. 

So there are two ways to fix this:
1. We can use the code you have added here and it will make sure that the new subscribed attributes in the sub driver are checked and added to device's subscribed attribute list when `info_changed` is run.
2. We expand the `subscribed_attributes` list in the parent driver to include the new attributes that are in this driver. I think this is the better long term solution because the `device_subscribe()` call utilizes the `subscribed_attributes` defined in the parent driver to determine which attributes a device should subscribe to. Therefore, the parent driver's `subscribed_attributes` list should be the "master list" for all attributes that could be subscribed to in the driver, including in the sub drivers. Therefore, all you would need to do is combine the subscribed attributes and subscribed events lists in the sub driver with the lists that already exist in the parent driver. This would look like this:
```
subscribed_attributes = {
     -- attributes already defined in parent driver
    [capabilities.lock.ID] = {
      DoorLock.attributes.LockState
    },
    [capabilities.battery.ID] = {
      PowerSource.attributes.BatPercentRemaining
    },
    [capabilities.batteryLevel.ID] = {
      PowerSource.attributes.BatChargeLevel
    },
    -- new attributes from new-matter-lock subdriver
    [capabilities.remoteControlStatus.ID] = {
      DoorLock.attributes.OperatingMode
    },
    [capabilities.lockUsers.ID] = {
      DoorLock.attributes.NumberOfTotalUsersSupported
    },
    [capabilities.lockCredentials.ID] = {
      DoorLock.attributes.NumberOfPINUsersSupported,
      DoorLock.attributes.MaxPINCodeLength,
      DoorLock.attributes.MinPINCodeLength,
      DoorLock.attributes.RequirePINforRemoteOperation
    },
    [capabilities.lockSchedules.ID] = {
      DoorLock.attributes.NumberOfWeekDaySchedulesSupportedPerUser,
      DoorLock.attributes.NumberOfYearDaySchedulesSupportedPerUser
    }
  },
 subscribed_events = {
    -- events already defined in parent driver
    [capabilities.tamperAlert.ID] = {
      DoorLock.events.DoorLockAlarm,
      DoorLock.events.LockOperation
    },
    [capabilities.lockAlarm.ID] = {
      DoorLock.events.DoorLockAlarm
    },
    [capabilities.lockCodes.ID] = {
      DoorLock.events.LockUserChange
    },
    -- new events from new-matter-lock subdriver
    [capabilities.lock.ID] = {
      DoorLock.events.LockOperation
    },
    [capabilities.lockUsers.ID] = {
      DoorLock.events.LockUserChange
    }
  },
```

Either way should work in this case, but I think long term we should move to having all subscribed attributes and events listed in the parent driver so that they do not need to be redefined in the sub drivers. The parent driver's list would serve as a "master list" of all the subscribed attributes and events that could be used in that driver, and the sub drivers can inherit those attributes accordingly!

Reply:
Is there a way we can append to the parent driver's `subscribed_attributes` list from a subdriver? I'm thinking that the pattern for the parent driver to contain all attributes/events that a subdriver may be interested in isn't sustainable and can also be problematic if a subdriver only contains manufacturer specific cluster attributes/events that general devices driven from other subdrivers or parent one may not be able to service. 

Also to note that the `device::subscribe()` is idempotent if the same attributes/events are resubscribed so solution (1) should be fine even if it ends up running multiple times.

Reply:
I think then that it may be a good idea to add the new capabilities / attributes into the main driver's subscribed attributes table? That way, this behavior can be handled by `device:subscribe()`.

Reply:
However, even if you elect not to do this, I do believe that this check: `if device.profile.id ~= args.old_st_store.profile.id then` should be addded around the handling, since it is unrelated to the issue you mentioned and it stops us from attempting to resubscribe on every `info_changed` event.

Reply:
@samadDotDev we do not currently support the functionality you suggested, but I do think that's a good idea!

Reply:
> Is there a way we can append to the parent driver's subscribed_attributes list from a subdriver? 
Good question! I think this can be done in a few ways. We can add subscribed attributes directly to the driver's subscribed attributes list as done above through the `add_subscribed_attribute/event()` function mentioned above. Also, we could access the driver's subscribed attribute/event list directly through the devices table via `self.driver.subscribed_attributes` or `self.driver.subscribed_events`. This is already done [here](https://github.ecodesamsung.com/iot-hub/scripting-engine/blob/889c2977afe7a3b4821eddbf6907518a4ac9a7fb/lua_libs/st/matter/device.lua#L243). However, since this is in the device table this should only change the list on a device-by-device basis. At that point, I think just directly calling the `add_subscribed_attribute` call as needed would be easier.

FWIW, in the matter-appliance driver this style of including a "master list" in the parent driver is already in place: https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/40c1b690507068caf852ce0a82c36ca5d778f0c0/drivers/SmartThings/matter-appliance/src/init.lua#L59. Each device type is broken into its own sub driver with handlers for each of the relevant attributes implemented in the sub driver. Also, it should be noted that even if all these attributes are listed in on master list, only the attributes that match to a capability that the device has on it's profile will actually be subscribed to. For example, a refrigerator would completely skip over the attributes related to the `laundryWasherRiseMode` since it would not have that capability in its profile.

Also, [in our public docs is it specifically stated that the `subscribed_attributes` and `subscribed_events` are not supported in sub drivers](https://developer.smartthings.com/docs/edge-device-drivers/matter/driver.html#subscribed-events). I think this support could be added via a lua libs change to add in "official" support, but either of the options proposed above also work for our purposes. Even in the case of a manufacturer-specific cluster, we can just leave that cluster out of the main list and instead directly call `device:add_subscribed_attribute()` in the sub driver to set up the subscription when `device:subscribe()` is called. However, I think additional support for these cases in the lua libs could help prevent some of the inefficiency of sending subscirbe requests for attributes that don't exist on the device in the case that the device is using a capbality that is overloaded by multiple different mutually exclusive attributes, such as the `mode` cluster in `matter-appliance`. The subscribes still work correctly and only subscribe to the relevant attributes on the report, but adding support defining subscribed attributes/events specific to a sub driver could prevent small inconsistencies like this.

Reply:
To summarize, I think we can use the solution you already have here @HunsupJung . This way we can avoid adding the attributes/events to the parent driver as @samadDotDev  had suggested. However, I think you should still add `if device.profile.id ~= args.old_st_store.profile.id then` before calling subscribe since we wouldn't need to resubscribe unless there was a profile change.

Also, if we are going to use this solution rather than adding the attributes to the parent driver, then you will also need to add the code that adds the new subscribed_attributes and subscribed_events every time before `device:subscribe()` is called. Since subscribe is also called in `device_init() `on line 102, you will also need to add the same code there:
```
local function device_init(driver, device)
  device:set_component_to_endpoint_fn(component_to_endpoint)
  for cap_id, attributes in pairs(subscribed_attributes) do
    if device:supports_capability_by_id(cap_id) then
      for _, attr in ipairs(attributes) do
        device:add_subscribed_attribute(attr)
      end
    end
  end
  for cap_id, events in pairs(subscribed_events) do
    if device:supports_capability_by_id(cap_id) then
      for _, e in ipairs(events) do
        device:add_subscribed_event(e)
      end
    end
  end
  device:subscribe()
 end
```

I have created a ticket to add more official support for subscribed attributes in sub drivers here: https://smartthings.atlassian.net/browse/CHAD-14110

Reply:
I applied it. please confirm the updating